src/translate.py

```diff
@@ -1173,6 +1184,10 @@ def dependencies(self) -> List[Expression]:
         return [self.expr]
 
     def format(self, fmt: Formatter) -> str:
+        if self.expr.type.is_function():
+            # Functions are automatically converted to function pointers
+            # without an explicit `&` by the compiler
```

Comment:
Yeah - we handle it in `address_of_gsym`. (Although I think I want to try refactoring that and making `address_of_gsym` *always* return an `AddressOf`, and then doing the type detection here at format time? I think that would solve the `StringLiteral` problem you mentioned, and make it easier to distinguish arrays in `.rodata`/`.data`.)

Reply:
The `StringLiteral` logic I agree should be moved to formatting (though it may still be nice to annotating the type as char* early). Array->pointer decay I'm less sure about, it's not just a formatting concern since it's really changing the type. But if the decay can be moved into other parts of the type inference code (e.g. decaying whenever we do a dereference or arithmetic) it could work.